### PR TITLE
🌱 Add conditions to the DockerMachine object

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR adds conditions to the DockerMachine object

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/3111

### CODE REVIEW

File: test/infrastructure/docker/api/v1alpha3/dockermachine_types.go [hunk 1]
```diff
@@ -79,6 +80,14 @@ type DockerMachineStatus struct {
 	// added to the load balancer
 	// +optional
 	LoadBalancerConfigured bool `json:"loadBalancerConfigured,omitempty"`
+
+	// ImagePreLoaded denotes that images where pre-loaded into the machine
+	// +optional
+	ImagePreLoaded bool `json:"imagePreLoaded,omitempty"`
```
Comment (fabriziopandini) [inline test/infrastructure/docker/api/v1alpha3/dockermachine_types.go:86]: This change is not strictly related to this PR, but it helps to get the code re-entrant in case of problems during container provisioning

File: test/infrastructure/docker/controllers/dockermachine_controller.go [hunk 2]
```diff
@@ -185,18 +192,34 @@ func (r *DockerMachineReconciler) reconcileNormal(ctx context.Context, machine *
...
-		return ctrl.Result{}, errors.Wrap(err, "failed to create worker DockerMachine")
+	// Create the machine if not existing yet
+	if !externalMachine.Exists() {
+		if err := externalMachine.Create(ctx, role, machine.Spec.Version, dockerMachine.Spec.ExtraMounts); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to create worker DockerMachine")
+		}
+	}
+
+	// Update the ContainerProvisionedCondition condition
+	if externalMachine.Exists() {
```
Comment (fabriziopandini) [inline test/infrastructure/docker/controllers/dockermachine_controller.go:213]: Nb. I'm implementing conditions.MarkTrue in a separate code branch vs the `if !externalMachine.Exists() {` branch above because I want to make the code re-entrant (vs settin condition true only after externalMachine.Create succeeded)

File: test/infrastructure/docker/controllers/dockermachine_controller.go [hunk 3]
```diff
@@ -209,34 +232,48 @@ func (r *DockerMachineReconciler) reconcileNormal(ctx context.Context, machine *
...
 		if err := externalMachine.ExecBootstrap(timeoutctx, bootstrapData); err != nil {
-			return ctrl.Result{}, errors.Wrap(err, "failed to exec DockerMachine bootstrap")
+			conditions.MarkFalse(dockerMachine, infrav1.BootstrapExecSucceededCondition, infrav1.BootstrapFailedReason, clusterv1.ConditionSeverityWarning, "Repeating bootstrap")
+			return ctrl.Result{RequeueAfter: 5 * time.Second}, errors.Wrap(err, "failed to exec DockerMachine bootstrap")
 		}
 		dockerMachine.Spec.Bootstrapped = true
 	}
 
+	// Update the BootstrapExecSucceededCondition condition
+	if dockerMachine.Spec.Bootstrapped {
```
Comment (fabriziopandini) [inline test/infrastructure/docker/controllers/dockermachine_controller.go:261]: Same as above (using a separated code branch to get the code re-entrant vs setting on ExecBootstrap succeded only)

File: test/infrastructure/docker/controllers/dockermachine_controller.go [hunk 4]
```diff
@@ -209,34 +232,48 @@ func (r *DockerMachineReconciler) reconcileNormal(ctx context.Context, machine *
 		dockerMachine.Status.LoadBalancerConfigured = true
 	}
 
+	// At, this stage, we are ready for bootstrap. However, if the BootstrapExecSucceededCondition is missing we add it and we
+	// requeue so the user can se the change of state before the bootstrap actually start.
+	if !conditions.Has(dockerMachine, infrav1.BootstrapExecSucceededCondition) {
```
Comment (detiber) [inline test/infrastructure/docker/controllers/dockermachine_controller.go:237]: I don't think we should be relying on conditions from previous reconciliations here, since we are the controller setting the conditions. Can we switch to detecting this through introspection instead?

File: test/infrastructure/docker/controllers/dockermachine_controller.go (hunk 3 above)
Comment (detiber) [reply test/infrastructure/docker/controllers/dockermachine_controller.go:261]: I don't think we should be using the conditions as a mechanism for re-entrancy. They exist in status, so should be re-create-able.

File: test/infrastructure/docker/controllers/dockermachine_controller.go [hunk 5]
```diff
@@ -185,18 +192,34 @@ func (r *DockerMachineReconciler) reconcileNormal(ctx context.Context, machine *
...
+
+	// Update the ContainerProvisionedCondition condition
+	if externalMachine.Exists() {
+		conditions.MarkTrue(dockerMachine, infrav1.ContainerProvisionedCondition)
 	}
 
 	// Preload images into the container
-	if err := externalMachine.PreloadLoadImages(ctx, dockerMachine.Spec.PreLoadImages); err != nil {
-		return ctrl.Result{}, errors.Wrap(err, "failed to pre-load images into the DockerMachine")
+	if !dockerMachine.Status.ImagePreLoaded && len(dockerMachine.Spec.PreLoadImages) > 0 {
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply test/infrastructure/docker/controllers/dockermachine_controller.go:218]: This was already the case for `LoadBalancerConfigured`, and I went through the same path considering this less flaky than doing inspection with docker exec.
but I agree that removing this two flags could be an interesting follow-up PR if we get docker exec to work reliably (I observed many flakes on CI, especially on newly created nodes)

File: test/infrastructure/docker/controllers/dockermachine_controller.go [hunk 6]
```diff
@@ -209,34 +232,48 @@ func (r *DockerMachineReconciler) reconcileNormal(ctx context.Context, machine *
 		dockerMachine.Status.LoadBalancerConfigured = true
 	}
 
+	// At, this stage, we are ready for bootstrap. However, if the BootstrapExecSucceededCondition is missing we add it and we
+	// requeue so the user can se the change of state before the bootstrap actually start.
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply test/infrastructure/docker/controllers/dockermachine_controller.go:236]: You are right, the main reason for this is that it makes a nicer UX when working with CAPD, because this is where the docker-machine provisioning spends most of the time.

File: test/infrastructure/docker/controllers/dockermachine_controller.go [hunk 7]
```diff
@@ -209,34 +232,48 @@ func (r *DockerMachineReconciler) reconcileNormal(ctx context.Context, machine *
...
+		return ctrl.Result{Requeue: true}, nil
+	}
+
 	// if the machine isn't bootstrapped, only then run bootstrap scripts
 	if !dockerMachine.Spec.Bootstrapped {
 		bootstrapData, err := r.getBootstrapData(ctx, machine)
 		if err != nil {
 			r.Log.Error(err, "failed to get bootstrap data")
-			return ctrl.Result{}, nil
+			return ctrl.Result{}, err
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply test/infrastructure/docker/controllers/dockermachine_controller.go:247]: Without any comments on why this was shallowed before, my assumption is that this was a mistake. So I fixed it

File: test/infrastructure/docker/controllers/dockermachine_controller.go (hunk 3 above)
Comment (fabriziopandini) [reply test/infrastructure/docker/controllers/dockermachine_controller.go:261]: Sorry if my explanation created confusion, but we are not using conditions for re-entrancy here.
However, we are ensuring the condition get always set and not only during the reconciliation that executes the bootstrap code (condition is re-create-able)

File: test/infrastructure/docker/controllers/dockermachine_controller.go (hunk 4 above)
Comment (fabriziopandini) [reply test/infrastructure/docker/controllers/dockermachine_controller.go:237]: Uhh... this can potentially apply to any info/flag a controller is setting in status, not only on conditions, and we are consistently relying on that information on our controllers.

However, in this specific case, this is specifically for ensuring a condition is reported to users as explained in a previous comment, so I find it weird to use introspection when what I want to check is explicit if a condition exists

Comment (detiber) [reply test/infrastructure/docker/controllers/dockermachine_controller.go:237]: > Uhh... this can potentially apply to any info/flag a controller is setting in status, not only on conditions, and we are consistently relying on that information on our controllers.

We shouldn't be, unless it's a status value that we've previously set as part of the current reconciliation or a resource that is managed by another controller. We should likely file bugs for any places we are currently relying on status set on previous reconciliations.

File: test/infrastructure/docker/api/v1alpha3/dockermachine_types.go [hunk 8]
```diff
@@ -79,6 +80,14 @@ type DockerMachineStatus struct {
 	// added to the load balancer
 	// +optional
 	LoadBalancerConfigured bool `json:"loadBalancerConfigured,omitempty"`
+
+	// ImagePreLoaded denotes that images where pre-loaded into the machine
```
Comment (benmoss) [inline test/infrastructure/docker/api/v1alpha3/dockermachine_types.go:84]:
```suggestion
	// ImagePreLoaded denotes that images were pre-loaded into the machine
```

File: test/infrastructure/docker/controllers/dockermachine_controller.go [hunk 9]
```diff
@@ -114,28 +109,16 @@ func (r *DockerMachineReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
...
-
 	// Initialize the patch helper
 	patchHelper, err := patch.NewHelper(dockerMachine, r)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 	// Always attempt to Patch the DockerMachine object and status after each reconciliation.
 	defer func() {
+		// always update the readyCondition; the summary is represented using the "1 of x completed" notation.
+		conditions.SetSummary(dockerMachine, conditions.WithStepCounter(2))
```
Comment (benmoss) [inline test/infrastructure/docker/controllers/dockermachine_controller.go:120]: What does the 2 represent here? The number of conditions on the `dockerMachine`? Should this be a constant then?

File: test/infrastructure/docker/controllers/dockermachine_controller.go (hunk 2 above)
Comment (benmoss) [reply test/infrastructure/docker/controllers/dockermachine_controller.go:213]: It feels like this check is unnecessary though now, right? Like isn't it unreachable unless `externalMachine.Exists()`?

File: test/infrastructure/docker/controllers/dockermachine_controller.go (hunk 4 above)
Comment (detiber) [reply test/infrastructure/docker/controllers/dockermachine_controller.go:237]: Yes, I believe this use case is fine and makes sense, thanks :)

File: test/infrastructure/docker/controllers/dockermachine_controller.go (hunk 3 above)
Comment (benmoss) [reply test/infrastructure/docker/controllers/dockermachine_controller.go:261]: Same as my other comment though, isn't this already unreachable if `dockerMachine.Spec.Bootstrapped` isn't true?

File: test/infrastructure/docker/controllers/dockermachine_controller.go (hunk 6 above)
Comment (benmoss) [inline test/infrastructure/docker/controllers/dockermachine_controller.go:236]:
```suggestion
	// requeue so the user can see the change of state before the bootstrap actually starts.
```

File: test/infrastructure/docker/controllers/dockermachine_controller.go [hunk 10]
```diff
@@ -209,34 +232,48 @@ func (r *DockerMachineReconciler) reconcileNormal(ctx context.Context, machine *
...
 	}
 
+	// Update the BootstrapExecSucceededCondition condition
+	if dockerMachine.Spec.Bootstrapped {
+		conditions.MarkTrue(dockerMachine, infrav1.BootstrapExecSucceededCondition)
+	}
+
 	// Usually a cloud provider will do this, but there is no docker-cloud provider.
-	// Requeue after 1s if there is an error, as this is likely momentary load balancer
+	// Requeue after if there is an error, as this is likely momentary load balancer
```
Comment (benmoss) [inline test/infrastructure/docker/controllers/dockermachine_controller.go:266]:
```suggestion
	// Requeue if there is an error, as this is likely momentary load balancer
```

**Comment by vincepri** [comment]
/retitle 🏃 Add conditions to the DockerMachine object

**Comment by vincepri** [comment]
/retitle 🌱 Add conditions to the DockerMachine object

**Comment by fabriziopandini** [comment]
I'm going to remove ImageLoaded flag...

File: test/infrastructure/docker/controllers/dockermachine_controller.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply test/infrastructure/docker/controllers/dockermachine_controller.go]: I removed the ImagePreLoaded flag.

**Comment by fabriziopandini** [comment]
@detiber flag removed (last commit)
Let me know if I can squash commits or there is something else to be addressed

**Comment by detiber** [comment]
lgtm pending squash

**Comment by fabriziopandini** [comment]
@detiber re-based

**Comment by vincepri** [comment]
LGTM

File: test/infrastructure/docker/controllers/dockermachine_controller.go [hunk 11]
```diff
@@ -209,34 +228,48 @@ func (r *DockerMachineReconciler) reconcileNormal(ctx context.Context, machine *
...
+			return ctrl.Result{}, err
 		}
 
 		timeoutctx, cancel := context.WithTimeout(ctx, 3*time.Minute)
 		defer cancel()
 		// Run the bootstrap script. Simulates cloud-init.
 		if err := externalMachine.ExecBootstrap(timeoutctx, bootstrapData); err != nil {
-			return ctrl.Result{}, errors.Wrap(err, "failed to exec DockerMachine bootstrap")
+			conditions.MarkFalse(dockerMachine, infrav1.BootstrapExecSucceededCondition, infrav1.BootstrapFailedReason, clusterv1.ConditionSeverityWarning, "Repeating bootstrap")
+			return ctrl.Result{RequeueAfter: 5 * time.Second}, errors.Wrap(err, "failed to exec DockerMachine bootstrap")
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply test/infrastructure/docker/controllers/dockermachine_controller.go]: fixed

**Comment by fabriziopandini** [comment]
@detiber PTAL
